src/test/debugger/appWorker.test.ts

```diff
@@ -6,80 +6,68 @@ import * as WebSocket from "ws";
-                });
-            });
+                testWorker = new ForkedAppWorker(packagerPort, sourcesStoragePath, postReplyFunction);
+                return testWorker;
+            }
 
             teardown(function() {
                 // Reset everything
-                sandboxedWorker = null;
-
-                readFileStub = sinon.stub();
-                downloadAppScriptStub = sinon.stub();
-                postReplyFunction = sinon.stub();
+                spawnStub.restore();
+                postReplyFunction.reset();
+                if (testWorker) {
+                    testWorker.stop();
+                }
             });
 
-
             test("should execute scripts correctly and be able to invoke the callback", function() {
                 const expectedMessageResult = { success: true };
                 const startScriptContents = `var testResponse = ${JSON.stringify(expectedMessageResult)}; postMessage(testResponse);`;
 
-                readFileStub.withArgs(startScriptFileName).returns(Q.resolve(startScriptContents));
-
-                return sandboxedWorker.start().then(() => {
-                    assert(postReplyFunction.calledWithExactly(expectedMessageResult));
-                });
+                return workerWithScript(startScriptContents).start()
+                .then(() =>
+                    Q.delay(100))
+                .then(() =>
+                    assert(postReplyFunction.calledWithExactly(expectedMessageResult)));
             });
 
             test("should be able to import scripts", function() {
-                const scriptImportPath = "testScript.js";
+                // NOTE: we're not able to mock out
+                const scriptImportPath = path.resolve(sourcesStoragePath, "importScriptsTest.js");
                 const startScriptContents = `importScripts("${scriptImportPath}"); postMessage("postImport");`;
-                readFileStub.withArgs(startScriptFileName).returns(Q.resolve(startScriptContents));
-
-                const testScriptContents = "postMessage('inImport')";
-                const scriptImportDeferred = Q.defer<void>();
-                downloadAppScriptStub.withArgs(scriptImportPath, debugAdapterPort).returns(scriptImportDeferred.promise.then(() => {
-                    return {
-                        contents: testScriptContents,
-                        filepath: scriptImportPath,
-                    };
-                }));
-
-                return sandboxedWorker.start().then(() => {
+
+                return workerWithScript(startScriptContents).start().then(() => {
                     // We have not yet finished importing the script, we should not have posted a response yet
                     assert(postReplyFunction.notCalled, "postReplyFuncton called before scripts imported");
-                    scriptImportDeferred.resolve(void 0);
-                    return Q.delay(1);
+                    return Q.delay(100);

```

Comment:
I think that previously the timeout was required just to delay dispatching the current promise chain and let `scriptImportDeferred` chain to be dispatched first.

In this implementation the spawned process is a black box and completely async for us so we can't reliably determine when the script has been imported and executed, so the only way to test this IMO is to give some reasonable amount of time to worker to complete script's import before checking if it has been done.

As for race conditions - not sure where are they might happen. Every test spawns it's own worker, so they are isolated from each other.